Hashtable.c

```diff
@@ -171,87 +185,120 @@ void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
 
    assert(Hashtable_isConsistent(this));
    assert(this->size > 0);
+   assert(value);
 
-   // grow on load > 1.0
-   if (this->items > this->size)
+   // grow on load > 0.7
+   if (10 * this->items > 7 * this->size)
       Hashtable_setSize(this, 2 * this->size);
 
-   unsigned int index = key % this->size;
-   HashtableItem** bucketPtr = &(this->buckets[index]);
-   while (true) {
-      if (*bucketPtr == NULL) {
-         *bucketPtr = HashtableItem_new(key, value);
-         this->items++;
-         break;
-      } else if ((*bucketPtr)->key == key) {
-         if (this->owner && (*bucketPtr)->value != value)
-            free((*bucketPtr)->value);
-         (*bucketPtr)->value = value;
-         break;
-      } else {
-         bucketPtr = &((*bucketPtr)->next);
-      }
-   }
+   insert(this, key, value);
 
    assert(Hashtable_isConsistent(this));
+   assert(Hashtable_get(this, key) != NULL);
+   assert(this->size > this->items);
 }
 
 void* Hashtable_remove(Hashtable* this, hkey_t key) {
    unsigned int index = key % this->size;
+   unsigned int origIndex = index;
+   unsigned int probe = 0;
 
    assert(Hashtable_isConsistent(this));
 
-   HashtableItem** bucket;
-   for (bucket = &(this->buckets[index]); *bucket; bucket = &((*bucket)->next) ) {
-      if ((*bucket)->key == key) {
-         void* value = (*bucket)->value;
-         HashtableItem* next = (*bucket)->next;
-         free(*bucket);
-         (*bucket) = next;
-         this->items--;
-         if (this->owner) {
-            free(value);
-            assert(Hashtable_isConsistent(this));
-            return NULL;
-         } else {
-            assert(Hashtable_isConsistent(this));
-            return value;
+   void* res = NULL;
+
+   while (this->buckets[index].value) {
+      if (this->buckets[index].key == key) {
+         res = this->owner ? (free(this->buckets[index].value), NULL) : this->buckets[index].value;
```

Comment:
Please write a full if statement. This is hard to read (although I know what you mean).